File: 5. Hash Table/Using Open Addressing/hashTable.js
const prompt = require('prompt-sync')({ sigint: true });

let TABLE_SIZE = 10;
let valueHash = [];
let keyHash = [];

function hashFunction(key) {
    return key % TABLE_SIZE;
}

function insert() {
    console.log("");
    let key = Number.parseInt(prompt("Enter Key:- "));

    if (key < 1) {
        console.log("\nInvalid Key!");
        return;
    }

    let value = Number.parseInt(prompt("Enter Value:- "));

    let index = hashFunction(key);
    let originalIndex = index;

    while (true) {
        if (keyHash[index] == key) {
            console.log("\nDuplicate Key Already Exists!\n1. Update\n2. Skip");
            let choice = Number.parseInt(prompt("Enter Your Choice:- "));
            if (choice == 1) {
                valueHash[index] = value;
                console.log("\nUpdate Successful!");
            }
            return;
        }
        index = (index + 1) % TABLE_SIZE;
        if (index == originalIndex) {
            break;
        }
    }

    while (keyHash[index]) {
        index = (index + 1) % TABLE_SIZE;
        if (index == originalIndex) {
            console.log("\nHash table is full!");
            return;
        }
    }

    keyHash[index] = key;
    valueHash[index] = value;

    console.log("\nInsertion Successful!");
}

function search() {
    console.log("");
    let key = Number.parseInt(prompt("Enter Key:- "));

    if (key < 1) {
        console.log("\nInvalid Key!");
        return;
    }

    let index = hashFunction(key);
    let originalIndex = index;

    while (true) {
        if (keyHash[index] == key) {
            console.log("Value: " + valueHash[index]);
            return;
        }
        index = (index + 1) % TABLE_SIZE;
        if (index == originalIndex) {
            break;
        }
    }

    console.log("\n~~ Key Not Found ~~");
}

function deleteKey() {
    console.log("");
    let key = Number.parseInt(prompt("Enter Key:- "));

    if (key < 1) {
        console.log("\nInvalid Key!");
        return;
    }

    let index = hashFunction(key);
    let originalIndex = index;

    while (true) {
        if (keyHash[index] == key) {
            valueHash[index] = null;
            keyHash[index] = null;
            console.log("Deletion Successful!");
            return;
        }
        index = (index + 1) % TABLE_SIZE;
        if (index == originalIndex) {
            break;
        }
    }

    console.log("\n~~ Key Not Found ~~");
}

function display() {
    console.log("");
    let str = "";
    for (let i = 0; i < TABLE_SIZE; i++) {
        if(keyHash[i]) {
            str = str + "(" + keyHash[i] + ", " + valueHash[i] + ")";
        } else {
            str = str + "NULL";
        }

        if (i != TABLE_SIZE - 1) {
            str = str +  " , ";
        }
    }
    console.log(str);
}

(() => {
    let choice;
    console.log("Hello! Here You Can Perform Following Hash Table/Maps Operations!");
    while (1) {
        console.log("\n1. Show Data\n2. Insert\n3. Delete With Key\n4. Check Value at a Key\n5. Exit");
        choice = Number.parseInt(prompt("Enter Your Choice:- "));
        switch (choice) {
            case 1:
                display();
                break;
            case 2:
                insert();
                break;
            case 3:
                deleteKey();
                break;
            case 4:
                search();
                break;
            case 5:
                console.log("\n~~ Thanks for Using! Have a Great Day! ~~");
                return;
            default:
                console.log("\nPlease Enter a Valid Input!");
        }
    }
    return;
})();

File: 5. Hash Table/Using Open Addressing/HashTable.java
import java.util.Scanner;

public class HashTable {
    private static final int EMPTY = -1;
    private static final int TABLE_SIZE = 10;
    private int[] valueHash;
    private int[] keyHash;

    public HashTable() {
        valueHash = new int[TABLE_SIZE];
        keyHash = new int[TABLE_SIZE];

        for (int i = 0; i < TABLE_SIZE; i++) {
            keyHash[i] = -1;
            valueHash[i] = -1;
        }
    }

    private int hashFunction(int key) {
        return key % TABLE_SIZE;
    }

    public void insert(Scanner scanner) {
        System.out.print("\nEnter Key: ");
        int key = scanner.nextInt();

        if (key < 1) {
            System.out.println("\nInvalid Key!");
            return;
        }

        System.out.print("Enter Value: ");
        int value = scanner.nextInt();

        int index = hashFunction(key);
        int originalIndex = index;

        while (true) {
            if (keyHash[index] == key) {
                System.out.print("\nDuplicate Key Already Exists!\n1. Update\n2. Skip\nEnter Choice: ");
                int choice = scanner.nextInt();
                if (choice == 1) {
                    valueHash[index] = value;
                    System.out.println("\nUpdate Successful!");
                }
                return;
            }
            index = (index + 1) % TABLE_SIZE;
            if (index == originalIndex) {
                break;
            }
        }

        while (keyHash[index] != EMPTY) {
            index = (index + 1) % TABLE_SIZE;
            if (index == originalIndex) {
                System.out.print("\nHash table is full!\n");
                return;
            }
        }

        keyHash[index] = key;
        valueHash[index] = value;

        System.out.println("\nInsertion Successful!");
    }

    public void search(Scanner scanner) {
        System.out.print("\nEnter Key: ");
        int key = scanner.nextInt();

        if (key < 1) {
            System.out.println("\nInvalid Key!");
            return;
        }

        int index = hashFunction(key);
        int originalIndex = index;

        while (true) {
            if (keyHash[index] == key) {
                System.out.println("Value: " + valueHash[index]);
                return;
            }
            index = (index + 1) % TABLE_SIZE;
            if (index == originalIndex) {
                break;
            }
        }

        System.out.println("\n~~ Key Not Found ~~");
    }
    
    public void delete(Scanner scanner) {
        System.out.print("\nEnter Key: ");
        int key = scanner.nextInt();
        
        if (key < 1) {
            System.out.println("\nInvalid Key!");
            return;
        }
        
        int index = hashFunction(key);
        int originalIndex = index;
        
        while (true) {
            if (keyHash[index] == key) {
                valueHash[index] = -1;
                keyHash[index] = -1;
                System.out.println("Deletion Successful!");
                return;
            }
            index = (index + 1) % TABLE_SIZE;
            if (index == originalIndex) {
                break;
            }
        }

        System.out.println("\n~~ Key Not Found ~~");
    }

    public void display() {
        System.out.print("\n");
        for (int i = 0; i < TABLE_SIZE; i++) {
            if(keyHash[i] != EMPTY) {
                System.out.format("(%d, %d)", keyHash[i], valueHash[i]);
            } else {
                System.out.print("NULL");
            }
    
            if (i != TABLE_SIZE - 1) {
                System.out.print(" , ");
            }
        }
        System.out.print("\n");
        return;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        HashTable hash = new HashTable();
        int choice;

        System.out.println("Hello! Here You Can Perform Following Hash Table/Maps Operations!");
        while (true) {
            System.out.println("\n1. Show Data\n2. Insert\n3. Delete With Key\n4. Check Value at a Key\n5. Exit");
            System.out.print("Enter Your Choice: ");
            choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    hash.display();
                    break;
                case 2:
                    hash.insert(scanner);
                    break;
                case 3:
                    hash.delete(scanner);
                    break;
                case 4:
                    hash.search(scanner);
                    break;
                case 5:
                    System.out.println("\n~~ Thanks for Using! Have a Great Day! ~~");
                    scanner.close();
                    return;
                default:
                    System.out.println("\nPlease Enter a Valid Input!");
            }
        }
    }
}
